### PR TITLE
[IR] Implementation of QoS Features

### PR DESCRIPTION
**Pull request by mm318**
Implementation of QoS features in the rclcpp package.

### CODE REVIEW

File: rclcpp/include/rclcpp/node.hpp [hunk 1]
```diff
@@ -142,76 +142,27 @@ class Node : public std::enable_shared_from_this<Node>
   /// Create and return a Publisher.
   /**
    * \param[in] topic_name The topic for this publisher to publish on.
-   * \param[in] qos_history_depth The depth of the publisher message queue.
```
Comment (emersonknapp) [inline rclcpp/include/rclcpp/node.hpp:145]: Reminder: we need to add back in the pre-existing `create_publisher` and `create_subscriber` APIs, and to mark them as deprecated

File: rclcpp/include/rclcpp/publisher_options.hpp [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+  PublisherOptions &
+  liveliness_callback(const QOSLivelinessEventCallbackType & callback)
+  {
+    callbacks_.liveliness_callback_ = callback;
+    return *this;
+  }
+
+  RCLCPP_PUBLIC
+  const QOSLifespanEventCallbackType &
+  lifespan_callback() const
```
Comment (emersonknapp) [inline rclcpp/include/rclcpp/publisher_options.hpp:153]: Reminder: remove all instance of `lifespan_callback`

**Comment by emersonknapp** [comment]
Whoops commenting on the wrong review, I'll close this one to reduce confusion

**Comment by mm318** [comment]
The other review (https://github.com/aws-robotics/rclcpp/pull/2) is `qos-min` vs. `master`, but maybe it should have been `qos-min` vs `qos_reviewed`.

**Comment by emersonknapp** [comment]
Changed the base on qos-min

**Pull request closed without merging**